### ISSUE TITLE
Missing null terminator with scp causes non-zero server side return code

### ISSUE THREAD

**Issue by dansebcar** [curl/curl#5507]
### I did this

`curl --upload-file /some/file --user my-user:my-pass scp://127.0.0.1:8022/tmp`

Where I have an sshd with a minimal ForceCommand running on port 8022 as described here: https://github.com/dansebcar/curl-scp (the ForceCommand wraps the usual server side `scp` call in a Python script, with some logic based on the return code).

The wrapped server side scp subprocess returns 1 when I use curl on the client side. When I instead use scp on the client side, the server side process returns 0 because its stdin is null-terminated.

In both cases, the file is copied successfully, the only important differences are the return code and the null terminator.

For my use case I'm writing a C program which links to libcurl, so I'm able to work around this discrepancy by manually appending a null byte to the buffer (and incrementing the return value) in CURLOPT_READFUNCTION (so I actually end up reading CURLOPT_INFILESIZE + 1 bytes).

I originally wrote about this issue on the mailing list before I identified the underlying cause and workaround but didn't get any replies https://curl.haxx.se/mail/lib-2020-05/0002.html .

### I expected the following

I expected the two commands to behave in the same way, ie that curl would also write a trailing null byte to the server side scp subprocess' stdin.

Of course, a null terminator is useless because the stdin's first line already contains the length, and the subsequent data can contain nulls. However, it does seem to be expected for a successful return code.

### curl/libcurl version

curl/stable,stable,now 7.64.0-4+deb10u1 amd64 
libcurl4/stable,stable,now 7.64.0-4+deb10u1 amd64 

[curl -V output]

```
curl 7.64.0 (x86_64-pc-linux-gnu) libcurl/7.64.0 OpenSSL/1.1.1d zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5) libssh2/1.8.0 nghttp2/1.36.0 librtmp/2.3
Release-Date: 2019-02-06
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL
```

### operating system

debian buster

Linux dc 4.19.0-9-amd64 #1 SMP Debian 4.19.118-2 (2020-04-29) x86_64 GNU/Linux

**Comment by bagder** [comment]
Isn't it strange that nobody else ever reported this and that the test suite runs fine and transfer SCP fine without any extra null byte?

To me, this sounds like you have either an odd server side.

**Comment by dansebcar** [comment]
The exit code in question is normally discarded and invisible to all parties. It is unusual that I am inspecting this exit code, but my hope is that it would accurately report the success of the scp operation, which it currently doesn't do if I use curl.

Apart from the exit code, there's no indication either client/ server side that anything went wrong, so I'm not surprised this discrepancy has gone unnoticed.

To demonstrate that there's nothing odd about my server side setup, I created a [separate branch](https://github.com/dansebcar/curl-scp/tree/without-force) in which I create a `/usr/local/bin/scp` shell script which simply wraps the real `/usr/bin/scp` and logs the exit code, I can reproduce this issue with an otherwise totally vanilla sshd installation and configuration.

**Comment by bagder** [comment]
> The exit code in question is normally discarded and invisible to all parties.

So how come it is relevant in your case? Also, libssh2 is the underlying library that does the SCP transfer for us here so if there's a problem, I would presume the problem is actually in that library and not in curl code...  I realize this is unfortunate and I don't like to point fingers, but it is good to talk to the right people. It would be useful to modify one of libssh2's SCP examples to reproduce the problem.

**Comment by dansebcar** [comment]
It is relevant for me because I want to take different actions with my ForceCommand script if scp failed for any reason, but if I see exit code 1 I don't know if something actually went wrong or if my client is just using curl/ libssh2.

I don't know what the appropriate solution is. I don't know if the trailing null is required in the scp protocol. One could also argue that the fault is with my server's scp itself (libssh?) for even returning 1 in the first place, since there's no obvious reason to demand a trailing null. This is especially true if it's not a documented requirement of the protocol.

I'm hesitant to do any further digging, since I already have a workaround that I'm satisfied with for libcurl, and I unfortunately don't have the time to dig into the libssh2 examples.

**Comment by bagder** [comment]
I ran the `scp_write` example from libssh2 (using libssh2 from current git master) with enabled tracing and uploaded a file over SCP (the `README.md` file actually) and it reported no issues or problems when doing so even though it logs the exit code 1 you noticed:
~~~
[libssh2] 0.700217 Conn: Exit status 1 received for channel 0/0
~~~
So yes, this is a libssh2 issue (not a curl one). I will urge you to pursue this with the libssh2 team. 

The SCP part from the trace looks like below:
~~~
[libssh2] 0.691478 Transport: Packet type 91 received, length=17
[libssh2] 0.691486 Transport: Looking for packet of type: 91
[libssh2] 0.691494 Conn: Connection Established - ID: 0/0 win: 0/2097152 pack: 32768/32768
[libssh2] 0.691502 Conn: starting request(exec) on channel 0/0, message=scp -t '/tmp/foo'
=> libssh2_transport_write plain (18 bytes)
0000: 62 00 00 00 00 00 00 00  04 65 78 65 63 01 00 00 : b........exec...
0010: 00 11                                            : ..
=> libssh2_transport_write plain2 (17 bytes)
0000: 73 63 70 20 2D 74 20 27  2F 74 6D 70 2F 66 6F 6F : scp -t '/tmp/foo
0010: 27                                               : '
[libssh2] 0.691598 Socket: Sent 80/80 bytes at 0x55fe544a6468
=> libssh2_transport_write send() (80 bytes)
0000: 47 8E DF A4 8E B7 B3 C6  45 8E 01 D0 2A C0 A8 1E : G.......E...*...
0010: 22 4C 97 7F FE 9B 06 C6  DB 94 1F 83 DC 22 9B 1E : "L..........."..
0020: 3D C8 19 38 63 9D 3F 4F  C2 47 EB 23 21 B9 03 33 : =..8c.?O.G.#!..3
0030: 0A 86 E9 E3 2A 6C CA 15  CB 63 32 A4 17 D7 0A 02 : ....*l...c2.....
0040: 6E 22 D4 EE 4E 24 D0 5A  DB E6 7D 0A FF 0E 11 F2 : n"..N$.Z..}.....
[libssh2] 0.691643 Transport: Looking for packet of type: 99
[libssh2] 0.691650 Transport: Looking for packet of type: 100
[libssh2] 0.691660 Failure Event: -37 - Would block requesting SCP startup
[libssh2] 0.692242 Transport: Looking for packet of type: 99
[libssh2] 0.692251 Transport: Looking for packet of type: 100
[libssh2] 0.692259 Socket: Recved 112/16384 bytes to 0x55fe544a2428+0
=> libssh2_transport_read() raw (112 bytes)
0000: B8 00 92 87 A8 B4 CD DC  DA 93 06 95 00 8F 35 97 : ..............5.
0010: 70 42 EA F3 B4 4F 94 1F  AB DA DD 9A 7D B2 8E 7B : pB...O......}..{
0020: 0A EF 84 71 A8 8A 2F AC  00 4A 79 52 14 D9 02 CE : ...q../..JyR....
0030: BC AD C2 7C 25 84 C2 B4  3F 59 1A FD 6A BF B1 C6 : ...|%...?Y..j...
0040: 9F 6F 08 13 55 55 52 D2  1E 0F 56 49 57 DE 96 03 : .o..UUR...VIW...
0050: E0 60 FB DC 04 92 07 C2  D1 0D E3 08 20 73 14 ED : .`.......... s..
0060: 55 B0 C5 DE 42 A6 C0 F4  DE 6D 92 01 05 88 69 9B : U...B....m....i.
=> libssh2_transport_read() plain (9 bytes)
0000: 5D 00 00 00 00 00 20 00  00                      : ]..... ..
[libssh2] 0.692350 Transport: Packet type 93 received, length=9
[libssh2] 0.692356 Conn: Window adjust for channel 0/0, adding 2097152 bytes, new window_size=2097152
=> libssh2_transport_read() plain (5 bytes)
0000: 63 00 00 00 00                                   : c....
[libssh2] 0.692387 Transport: Packet type 99 received, length=5
[libssh2] 0.692396 Transport: Looking for packet of type: 99
[libssh2] 0.692404 Conn: channel_read() wants 1 bytes from channel 0/0 stream #0
[libssh2] 0.692415 Failure Event: -37 - would block
[libssh2] 0.692423 Failure Event: -37 - Would block waiting for response from remote
[libssh2] 0.699059 Conn: channel_read() wants 1 bytes from channel 0/0 stream #0
[libssh2] 0.699086 Socket: Recved 64/16384 bytes to 0x55fe544a2428+0
=> libssh2_transport_read() raw (64 bytes)
0000: 8B F9 F3 48 7E 46 CB 30  6C 95 47 2B FB E2 69 D6 : ...H~F.0l.G+..i.
0010: 67 2D 02 66 66 C0 EA 92  5D AA 7D A0 21 D4 9D DD : g-.ff...].}.!...
0020: F4 14 89 43 DD 60 99 9B  B2 49 F4 1B C9 A3 9B AE : ...C.`...I......
0030: 69 05 76 E2 43 22 FA C5  58 8C DD 94 97 AA AC C6 : i.v.C"..X.......
=> libssh2_transport_read() plain (10 bytes)
0000: 5E 00 00 00 00 00 00 00  01 00                   : ^.........
[libssh2] 0.699155 Transport: Packet type 94 received, length=10
[libssh2] 0.699160 Conn: 1 bytes packet_add() for 0/0/0
[libssh2] 0.699163 Conn: increasing read_avail by 1 bytes to 1/2097152
[libssh2] 0.699168 Conn: channel_read() got 1 of data from 0/0/0 [ul]
[libssh2] 0.699176 SCP: Sent C0644 442 foo

[libssh2] 0.699179 Conn: Writing 14 bytes on channel 0/0, stream #0
[libssh2] 0.699183 Conn: Sending 14 bytes on channel 0/0, stream_id=0
=> libssh2_transport_write plain (9 bytes)
0000: 5E 00 00 00 00 00 00 00  0E                      : ^........
=> libssh2_transport_write plain2 (14 bytes)
0000: 43 30 36 34 34 20 34 34  32 20 66 6F 6F 0A       : C0644 442 foo.
[libssh2] 0.699228 Socket: Sent 64/64 bytes at 0x55fe544a6468
=> libssh2_transport_write send() (64 bytes)
0000: B8 2D EE CC 1A 13 DA 40  A2 10 0D 17 D0 48 ED 13 : .-.....@.....H..
0010: 44 3A 82 DB 02 A9 C4 18  72 00 D1 9B B5 4B D3 81 : D:......r....K..
0020: 54 83 AD 76 5B BB 2D AA  D8 ED 9E B5 DD FE E2 AF : T..v[.-.........
0030: E3 CE FD BE 8C 27 2E F4  8B 75 6F AB 04 AC 02 80 : .....'...uo.....
[libssh2] 0.699250 Conn: channel_read() wants 1 bytes from channel 0/0 stream #0
[libssh2] 0.699255 Failure Event: -37 - would block
[libssh2] 0.699257 Failure Event: -37 - Would block waiting for response
[libssh2] 0.699362 Conn: channel_read() wants 1 bytes from channel 0/0 stream #0
[libssh2] 0.699376 Socket: Recved 64/16384 bytes to 0x55fe544a2428+0
=> libssh2_transport_read() raw (64 bytes)
0000: 91 ED BC 62 2E 34 86 EE  53 6B 03 A6 83 A4 B1 EA : ...b.4..Sk......
0010: 3E E3 CB FB ED F4 5A FC  8A D8 6C 79 E0 3D 0A 35 : >.....Z...ly.=.5
0020: F0 E2 85 5E C8 37 B9 BD  89 BF 95 B0 65 5C 4B 19 : ...^.7......e\K.
0030: 45 82 F3 29 CA C8 3D 9A  EC D2 DE 43 78 C9 CF BD : E..)..=....Cx...
=> libssh2_transport_read() plain (10 bytes)
0000: 5E 00 00 00 00 00 00 00  01 00                   : ^.........
[libssh2] 0.699405 Transport: Packet type 94 received, length=10
[libssh2] 0.699408 Conn: 1 bytes packet_add() for 0/0/0
[libssh2] 0.699411 Conn: increasing read_avail by 1 bytes to 1/2097151
[libssh2] 0.699415 Conn: channel_read() got 1 of data from 0/0/0 [ul]
SCP session waiting to send file
[libssh2] 0.699562 Conn: Writing 442 bytes on channel 0/0, stream #0
[libssh2] 0.699568 Conn: Sending 442 bytes on channel 0/0, stream_id=0
=> libssh2_transport_write plain (9 bytes)
0000: 5E 00 00 00 00 00 00 01  BA                      : ^........
=> libssh2_transport_write plain2 (442 bytes)
0000: 23 20 6C 69 62 73 73 68  32 20 2D 20 53 53 48 32 : # libssh2 - SSH2
0010: 20 6C 69 62 72 61 72 79  0A 0A 6C 69 62 73 73 68 :  library..libssh
0020: 32 20 69 73 20 61 20 6C  69 62 72 61 72 79 20 69 : 2 is a library i
0030: 6D 70 6C 65 6D 65 6E 74  69 6E 67 20 74 68 65 20 : mplementing the 
0040: 53 53 48 32 20 70 72 6F  74 6F 63 6F 6C 2C 20 61 : SSH2 protocol, a
0050: 76 61 69 6C 61 62 6C 65  20 75 6E 64 65 72 0A 74 : vailable under.t
0060: 68 65 20 72 65 76 69 73  65 64 20 42 53 44 20 6C : he revised BSD l
0070: 69 63 65 6E 73 65 2E 0A  0A 5B 57 65 62 20 73 69 : icense...[Web si
0080: 74 65 5D 28 68 74 74 70  73 3A 2F 2F 77 77 77 2E : te](https://www.
0090: 6C 69 62 73 73 68 32 2E  6F 72 67 2F 29 0A 0A 5B : libssh2.org/)..[
00a0: 4D 61 69 6C 69 6E 67 20  6C 69 73 74 5D 28 68 74 : Mailing list](ht
00b0: 74 70 73 3A 2F 2F 63 6F  6F 6C 2E 68 61 78 78 2E : tps://cool.haxx.
00c0: 73 65 2F 6D 61 69 6C 6D  61 6E 2F 6C 69 73 74 69 : se/mailman/listi
00d0: 6E 66 6F 2F 6C 69 62 73  73 68 32 2D 64 65 76 65 : nfo/libssh2-deve
00e0: 6C 29 0A 0A 5B 42 53 44  20 4C 69 63 65 6E 73 65 : l)..[BSD License
00f0: 64 5D 28 68 74 74 70 73  3A 2F 2F 6C 69 62 73 73 : d](https://libss
0100: 68 32 2E 6F 72 67 2F 6C  69 63 65 6E 73 65 2E 68 : h2.org/license.h
0110: 74 6D 6C 29 0A 0A 5B 57  65 62 20 73 69 74 65 20 : tml)..[Web site 
0120: 73 6F 75 72 63 65 20 63  6F 64 65 5D 28 68 74 74 : source code](htt
0130: 70 73 3A 2F 2F 67 69 74  68 75 62 2E 63 6F 6D 2F : ps://github.com/
0140: 6C 69 62 73 73 68 32 2F  77 77 77 29 0A 0A 49 6E : libssh2/www)..In
0150: 73 74 61 6C 6C 61 74 69  6F 6E 20 69 6E 73 74 72 : stallation instr
0160: 75 63 74 69 6F 6E 73 3A  0A 20 2D 20 5B 66 6F 72 : uctions:. - [for
0170: 20 43 4D 61 6B 65 5D 28  64 6F 63 73 2F 49 4E 53 :  CMake](docs/INS
0180: 54 41 4C 4C 5F 43 4D 41  4B 45 2E 6D 64 29 0A 20 : TALL_CMAKE.md). 
0190: 2D 20 5B 66 6F 72 20 61  75 74 6F 74 6F 6F 6C 73 : - [for autotools
01a0: 5D 28 64 6F 63 73 2F 49  4E 53 54 41 4C 4C 5F 41 : ](docs/INSTALL_A
01b0: 55 54 4F 54 4F 4F 4C 53  29 0A                   : UTOTOOLS).
[libssh2] 0.699668 Socket: Sent 496/496 bytes at 0x55fe544a6468
=> libssh2_transport_write send() (496 bytes)
0000: 05 D4 E1 ED 96 5D E8 3C  8C 45 E8 ED 1D 36 52 B2 : .....].<.E...6R.
0010: 71 95 B7 F4 4C CA D8 D1  C5 EB BF D0 7E CD 46 50 : q...L.......~.FP
0020: 96 0B 53 49 DD A2 F8 79  42 F8 A4 02 76 62 09 21 : ..SI...yB...vb.!
0030: 85 A9 71 EB B8 A3 A9 4F  4E 2A 55 AE 16 88 07 4E : ..q....ON*U....N
0040: A9 69 8B 1F 93 25 5E FD  02 F4 86 F3 F5 28 E6 BB : .i...%^......(..
0050: CC 7A 59 4A 30 7A 71 8A  80 52 30 81 85 0D A7 E6 : .zYJ0zq..R0.....
0060: 4F DA BF 15 A6 B4 91 43  AF 9B C2 B0 76 9F 4A 44 : O......C....v.JD
0070: 12 27 2B 3C 67 36 20 DF  1C 6F CB CA AA 44 FB 1E : .'+<g6 ..o...D..
0080: 26 D3 36 F9 8A E3 8E 4A  DE D0 85 7A D0 E4 2E 7B : &.6....J...z...{
0090: 33 99 54 CD 47 6D 4A EC  F2 4E FF 78 02 A6 E8 2C : 3.T.GmJ..N.x...,
00a0: D5 63 8A 67 76 94 8D 2B  BF 9A 28 4B B7 A9 D2 7B : .c.gv..+..(K...{
00b0: 49 DD 58 A5 31 4F 0C A1  35 F9 41 E7 8D 47 5A 0D : I.X.1O..5.A..GZ.
00c0: 88 02 B4 DA 43 5A 78 55  E7 6E E4 74 88 7C 17 0A : ....CZxU.n.t.|..
00d0: 3F 17 D6 DF B3 F3 81 D5  2A 05 35 E5 48 2E 98 9C : ?.......*.5.H...
00e0: 48 8C 9D DE 99 4A 9E 92  23 C6 8F 9A FD 03 F9 C8 : H....J..#.......
00f0: C0 68 E7 A4 EE 19 2D 0E  F8 D6 86 C2 F2 4B 7D 16 : .h....-......K}.
0100: 13 56 91 23 7E A8 42 6C  C5 1E FA FC 91 CD A1 50 : .V.#~.Bl.......P
0110: 43 04 D7 CE 02 61 C8 2F  36 0F 1E 16 6E 12 BD D4 : C....a./6...n...
0120: 47 A9 E0 F1 28 E4 FA 47  C3 BF 66 21 7E 2A 94 AC : G...(..G..f!~*..
0130: C7 13 F4 D8 9B AA 96 C1  D8 14 E9 0F E2 28 A3 41 : .............(.A
0140: 2F 90 22 32 CD B2 38 03  9A 2B FD 81 99 C1 BE 27 : /."2..8..+.....'
0150: 54 6D 72 21 92 74 B1 C1  23 BB 59 4F 31 BD 42 40 : Tmr!.t..#.YO1.B@
0160: A4 29 94 68 3A B2 1E 31  7A 22 03 AE A1 04 39 A9 : .).h:..1z"....9.
0170: A6 A8 40 49 9C 05 13 1B  5C 29 1B DD 75 7D 6C EF : ..@I....\)..u}l.
0180: 6E DA AA CE 9B EB E0 60  91 C9 95 0A 01 62 C4 55 : n......`.....b.U
0190: BE EE 07 70 9F 41 5A 47  92 A0 0D B4 A8 3C 64 BB : ...p.AZG.....<d.
01a0: C0 E9 65 32 C4 CD E9 AA  46 38 24 01 7C AF D7 BA : ..e2....F8$.|...
01b0: 9D ED 79 C2 CD 7E 52 ED  D0 DA 8E 97 B1 89 CA 21 : ..y..~R........!
01c0: C1 CE 97 B3 66 3A 34 23  2F 1D 0A BA CE B8 E5 5A : ....f:4#/......Z
01d0: 30 F0 47 E2 B6 AC A0 AB  1F B5 79 C2 0A 28 BE BA : 0.G.......y..(..
01e0: DD 22 F8 37 89 B3 F2 85  AD F5 B2 5B 2F 3F B4 89 : .".7.......[/?..
Sending EOF
[libssh2] 0.699779 Conn: Sending EOF on channel 0/0
=> libssh2_transport_write plain (5 bytes)
0000: 60 00 00 00 00                                   : `....
[libssh2] 0.699801 Socket: Sent 48/48 bytes at 0x55fe544a6468
=> libssh2_transport_write send() (48 bytes)
0000: 8D 83 EC B9 E4 C2 06 14  3E 5C 53 EB 55 94 D9 6C : ........>\S.U..l
0010: E8 6C 60 F4 35 37 A2 A5  CC 1D 7A F9 7A E4 75 42 : .l`.57....z.z.uB
0020: A4 8C B6 28 CE AC A3 67  34 F3 F9 F6 C6 E2 B5 92 : ...(...g4.......
Waiting for EOF
[libssh2] 0.699817 Conn: Awaiting EOF for channel 0/0
[libssh2] 0.700148 Socket: Recved 176/16384 bytes to 0x55fe544a2428+0
=> libssh2_transport_read() raw (176 bytes)
0000: 8D AC 50 85 3B 94 35 2C  64 59 B8 08 94 44 8A 27 : ..P.;.5,dY...D.'
0010: 79 CE B8 1E 50 11 17 11  6F 17 91 37 3B 18 1E 9C : y...P...o..7;...
0020: CF 99 59 ED A2 89 0C AF  9F 4B B2 72 E5 7E C6 6B : ..Y......K.r.~.k
0030: 17 7C 1C BB 43 DD 1D 6F  FC 04 08 5B 69 8E 32 1D : .|..C..o...[i.2.
0040: 06 1E 3B 2E 6D ED AC 20  91 D7 4F 88 45 D2 35 FD : ..;.m.. ..O.E.5.
0050: 06 41 1A 71 19 61 FD 40  29 26 D0 76 0D C2 CA 28 : .A.q.a.@)&.v...(
0060: D9 69 06 B9 95 4A 09 F1  67 FE C1 E8 69 38 EB CE : .i...J..g...i8..
0070: F3 D2 9A AE 7A 3B D7 79  25 87 97 19 2E 64 F7 64 : ....z;.y%....d.d
0080: 4A FF 4A F3 28 A0 70 CF  D2 F8 5D E2 7E E2 C2 7B : J.J.(.p...].~..{
0090: 75 D1 60 A4 68 B2 A5 CA  E6 CD 45 72 6A F0 4C 74 : u.`.h.....Erj.Lt
00a0: 1A F8 7A 30 F3 DC 25 E4  C0 95 3C 6E E8 74 75 72 : ..z0..%...<n.tur
=> libssh2_transport_read() plain (5 bytes)
0000: 60 00 00 00 00                                   : `....
[libssh2] 0.700189 Transport: Packet type 96 received, length=5
[libssh2] 0.700192 Conn: EOF received for channel 0/0
Waiting for channel to close
[libssh2] 0.700197 Conn: Awaiting close of channel 0/0
=> libssh2_transport_read() plain (25 bytes)
0000: 62 00 00 00 00 00 00 00  0B 65 78 69 74 2D 73 74 : b........exit-st
0010: 61 74 75 73 00 00 00 00  01                      : atus.....
[libssh2] 0.700211 Transport: Packet type 98 received, length=25
[libssh2] 0.700214 Conn: Channel 0 received request type exit-status (wr 0)
[libssh2] 0.700217 Conn: Exit status 1 received for channel 0/0
=> libssh2_transport_read() plain (5 bytes)
0000: 61 00 00 00 00                                   : a....
[libssh2] 0.700229 Transport: Packet type 97 received, length=5
[libssh2] 0.700231 Conn: Close received for channel 0/0
[libssh2] 0.700235 Conn: Freeing channel 0/0 resources
[libssh2] 0.700237 Conn: Closing channel 0/0
=> libssh2_transport_write plain (5 bytes)
0000: 61 00 00 00 00                                   : a....
[libssh2] 0.700286 Socket: Sent 48/48 bytes at 0x55fe544a6468
=> libssh2_transport_write send() (48 bytes)
0000: DB 7E C8 94 61 D7 F4 B5  AC 79 1B F4 1C CA 9C 7E : .~..a....y.....~
0010: 7A D0 A0 CB E3 DE 06 13  5D 18 04 92 BA CC D7 B2 : z.......].......
0020: C4 47 9D 16 90 DD 08 A1  5B 7D 90 2A CF 0F 4D F2 : .G......[}.*..M.
[libssh2] 0.700325 Transport: Looking for packet of type: 94
[libssh2] 0.700331 Transport: Looking for packet of type: 95
[libssh2] 0.700336 Transport: Disconnecting: reason=11, desc=Normal Shutdown, lang=
=> libssh2_transport_write plain (28 bytes)
0000: 01 00 00 00 0B 00 00 00  0F 4E 6F 72 6D 61 6C 20 : .........Normal 
0010: 53 68 75 74 64 6F 77 6E  00 00 00 00             : Shutdown....
=> libssh2_transport_write plain2 (0 bytes)
[libssh2] 0.700379 Socket: Sent 80/80 bytes at 0x55fe544a6468
=> libssh2_transport_write send() (80 bytes)
0000: EA F7 94 08 0D D8 87 70  FF 8D 50 B1 14 E3 5D AE : .......p..P...].
0010: 11 89 3B 6E B4 39 4D 0C  16 D4 A7 A2 B4 6F DC 45 : ..;n.9M......o.E
0020: 6C 0D 98 4D 51 BA 87 24  57 2A 31 87 2A B2 DF 88 : l..MQ..$W*1.*...
0030: 80 7B BE 2A CE 46 0A 72  06 D7 7E 23 35 5D 92 93 : .{.*.F.r..~#5]..
0040: 20 F4 93 E0 FB E4 82 E4  48 8A EB 8C 2E BB 0C A8 :  .......H.......
[libssh2] 0.700396 Transport: Freeing session resource
[libssh2] 0.700407 Transport: Extra packets left 0
all done
